templates/federated_reporting/mrproper.py

```diff
@@ -0,0 +1,244 @@
+
+    response = api.fr_remote_hubs()
+    for hub in response['hubs']:
+        feeder_credentials = getpass(prompt="Enter admin credentials for {} at {}".format(hub['ui_name'], hub['api_url']))
+        feederHostname = hub["ui_name"]
+        feederApi = NovaApi(
+            api_user="admin",
+            api_password=feeder_credentials,
+            cert_path=CERT_PATH,
+            hostname=feederHostname,
+        )
+
+        print("Creating mrproper role on {}".format(feederHostname))
+        response = feederApi.put(
+            "role",
+            "mrproper",
+            {
+                "description": "mrproper Federated Host Cleanup role",
+                "includeContext": "cfengine",
+            },
+        )
+        if response["status"] != 201:
+            print("Problem creating mrproper role on superhub. {}".format(response))
+            sys.exit(1)
+        response = feederApi.put_role_permissions("mrproper", ["host.delete"])
+        if response["status"] != 201:
+            print("Unable to set RBAC permissions on role mrproper")
+            sys.exit(1)
+        print("Creating mrproper user on {}".format(feederHostname))
+        response = feederApi.put(
+            "user",
+            "mrproper",
+            {
+                "description": "mrproper Federated Host Cleanup user",
+                "email": "{}".format(email),
+                "password": "{}".format(mrproper_password),
+                "roles": ["mrproper"],
+            },
+        )
+        if response["status"] != 201:
+            print("Problem creating mrproper user on {}. {}".format(feederHostname, response))
+            sys.exit(1)
+        write_secret(MRPROPER_SECRET_PATH, mrproper_password)
+
+
+def main():
+    if not os.path.exists(MRPROPER_SECRET_PATH):
+        if sys.stdout.isatty():
+            interactive_setup()
+        else:
+            print(
+                "{} requires manual setup, please run as root interactively.".format(
+                    os.path.basename(__file__)
+                )
+            )
+            sys.exit(1)
+
+    mrproper_password = read_secret(MRPROPER_SECRET_PATH)
+    api = NovaApi(
+        api_user="mrproper", api_password=mrproper_password
+    )  # defaults to localhost
+    response = api.fr_hub_status()
+    if not (
+        response["status"] == 200
+        and response["role"] == "superhub"
+        and response["configured"]
+    ):
+        print(
+            "{} can only be run on a Federated Reporting hub configured to be superhub".format(
+                os.path.basename(__file__)
+            )
+        )
+        sys.exit(1)
+
+    response = api.fr_remote_hubs()
+    for hub in response['hubs']:
+        if hub["role"] != "feeder":
+            continue
+
+        feeder_hostkey = hub["hostkey"]
+
+        sql = """
+SELECT DISTINCT hostkey
+FROM hosts
+WHERE EXISTS(
+  SELECT 1 FROM __hubs WHERE hostkey = '{}'
+  AND __hubs.hub_id = hosts.hub_id
+) AND EXISTS(
```

Comment:
What's the purpose of this first `EXISTS()` check? Every host belongs to some hub, right?

Reply:
This first one is to select the hostkeys associated with the feeder we are currently operating on.

Reply:
That's usually done with `WHERE ...` or `JOIN ON`. As written above, the block says "if there exists a hub with the given `hostkey` and a matching `hub_id`".

Reply:
Fixed this up.